src/blockcipher/aes_ctr.rs

```diff
@@ -0,0 +1,320 @@
+                    .clone()
+                    .verify(&self.exp_hmac)
+                    .is_err()
+            {
+                return Err(std::io::Error::new(
+                    std::io::ErrorKind::Other,
+                    format!(
+                        "failed decrypt byte stream, exp hmac: {:?} , actual hmac: {:?}",
+                        &self.exp_hmac,
+                        self.hmac.as_ref().unwrap().clone().finalize().into_bytes()
+                    ),
+                ));
+            }
+        }
+
+        self.cipher
+            .as_mut()
+            .unwrap()
+            .apply_keystream(&mut buf[0..read_len]);
+
+        if self.encrypt {
+            self.hmac.as_mut().unwrap().update(&buf[0..read_len]);
+
+            if self.done_encrypting {
+                self.exp_hmac = self
+                    .hmac
+                    .as_ref()
+                    .unwrap()
+                    .clone()
+                    .finalize()
+                    .into_bytes()
+                    .to_vec();
+            }
+        }
+
+        Ok(read_len)
+    }
+}
+
+impl<R: Read> AESCTRBlockCipher<R> {
+    pub fn new(bits: usize) -> Result<AESCTRBlockCipher<R>> {
+        if bits != AES256_KEY_SIZE * 8 {
+            return Err(anyhow!("AES CTR bit count not supported"));
+        }
+
+        Ok(AESCTRBlockCipher {
+            key_len: AES256_KEY_SIZE,
+            reader: None,
+            encrypt: false,
+            cipher: None,
+            hmac: None,
+            exp_hmac: vec![],
+            done_encrypting: false,
+        })
+    }
+
+    // init initializes an instance
+    pub fn init(
+        &mut self,
+        encrypt: bool,
+        reader: R,
+        opts: &mut LayerBlockCipherOptions,
+    ) -> Result<()> {
+        let symmetric_key = &opts.private.symmetric_key;
+        if symmetric_key.len() != AES256_KEY_SIZE {
+            return Err(anyhow!(
+                "invalid key length of {} bytes; need {} bytes",
+                symmetric_key.len(),
+                AES256_KEY_SIZE
+            ));
+        }
+
+        let mut nonce = vec![0u8; AES256_NONCE_SIZE];
+        match opts.get_opt("nonce".to_string()) {
+            Some(v) => {
+                if v.len() != AES256_NONCE_SIZE {
+                    return Err(anyhow!(
+                        "invalid nonce length of {} bytes; need {} bytes",
+                        nonce.len(),
+                        AES256_NONCE_SIZE
+                    ));
+                }
+                nonce = v;
+            }
+            None => {
+                thread_rng().try_fill(&mut nonce[..])?;
+            }
+        }
+
+        let cipher = Aes256Ctr::new(
+            GenericArray::from_slice(symmetric_key.as_slice()),
+            GenericArray::from_slice(nonce.as_slice()),
+        );
+
+        let hmac =
+            HmacSha256::new_from_slice(symmetric_key.as_slice()).expect("hmac use symmetric key");
```

Comment:
Should'nt we handle error here using match instead of expect, expect would result in panic right?

Reply:
hmac can take key of any size: https://docs.rs/hmac/0.11.0/hmac/ and will not panic , this expect is just compatiable with NewMac trait.

Reply:
ok

Reply:
@arronwy You're right, the Hmac implementation of `NewMac` does not panic. But this is an internal implementation and we should not rely on that. I agree with @pravinrajr9 we should handle the `Result` from `new_from_slice` and don't make any assumption on the internal implementation not panicking.

Reply:
After remove the Err wraper, it will works.